### PR TITLE
Add support for inferring the types of object arrays

### PR DESCRIPTION
**Pull request by eric-wieser**
This hopefully doesn't enable them to be used in any new jit contexts, but at least allows inspection of their type.

### CODE REVIEW

File: numba/np/numpy_support.py [hunk 1]
```diff
@@ -143,6 +145,8 @@ def as_dtype(nbtype):
     if isinstance(nbtype, types.NestedArray):
         spec = (as_dtype(nbtype.dtype), tuple(nbtype.shape))
         return np.dtype(spec)
+    if isinstance(nbtype, types.pyobject)
```
Comment (stuartarchibald) [inline numba/np/numpy_support.py]:
```suggestion
    if isinstance(nbtype, types.pyobject):
```

**Comment by stuartarchibald** [comment]
Thanks for the PR. Please could you add appropriate tests in `numba.tests.test_numpy_support`? Thanks.

**Comment by eric-wieser** [comment]
One small test added

**Comment by eric-wieser** [comment]
Tests seem to fail: https://dev.azure.com/numba/numba/_build/results?buildId=4565&view=logs&j=4f8e7448-6bbb-53dd-1446-49d51b72f33b&t=27cff68a-387b-5a51-18da-e9aa63580a90&l=311

```

    self.check_divmod_int(floordiv, [5, 0, 0, 2])
  File "/Users/runner/runners/2.165.0/work/1/s/numba/tests/npyufunc/test_errors.py", line 133, in check_divmod_int
    res = f(a, b)
  File "/Users/runner/runners/2.165.0/work/1/s/numba/np/ufunc/dufunc.py", line 190, in _compile_for_args
    return self._compile_for_argtys(tuple(argtys))
  File "/Users/runner/runners/2.165.0/work/1/s/numba/np/ufunc/dufunc.py", line 208, in _compile_for_argtys
    self._dispatcher, self.targetoptions, sig)
  File "/Users/runner/runners/2.165.0/work/1/s/numba/np/ufunc/ufuncbuilder.py", line 143, in _compile_element_wise_function
    cres = nb_func.compile(sig, **targetoptions)
  File "/Users/runner/runners/2.165.0/work/1/s/numba/np/ufunc/ufuncbuilder.py", line 92, in compile
    return self._compile_core(sig, flags, locals)
  File "/Users/runner/runners/2.165.0/work/1/s/numba/np/ufunc/ufuncbuilder.py", line 127, in _compile_core
    flags=flags, locals=locals)
  File "/Users/runner/runners/2.165.0/work/1/s/numba/core/compiler.py", line 549, in compile_extra
    return pipeline.compile_extra(func)
  File "/Users/runner/runners/2.165.0/work/1/s/numba/core/compiler.py", line 329, in compile_extra
    return self._compile_bytecode()
  File "/Users/runner/runners/2.165.0/work/1/s/numba/core/compiler.py", line 391, in _compile_bytecode
    return self._compile_core()
  File "/Users/runner/runners/2.165.0/work/1/s/numba/core/compiler.py", line 371, in _compile_core
    raise e
  File "/Users/runner/runners/2.165.0/work/1/s/numba/core/compiler.py", line 362, in _compile_core
    pm.run(self.state)
  File "/Users/runner/runners/2.165.0/work/1/s/numba/core/compiler_machinery.py", line 341, in run
    raise patched_exception
  File "/Users/runner/runners/2.165.0/work/1/s/numba/core/compiler_machinery.py", line 332, in run
    self._runPass(idx, pass_inst, state)
  File "/Users/runner/runners/2.165.0/work/1/s/numba/core/compiler_lock.py", line 32, in _acquire_compile_lock
    return func(*args, **kwargs)
  File "/Users/runner/runners/2.165.0/work/1/s/numba/core/compiler_machinery.py", line 291, in _runPass
    mutated |= check(pss.run_pass, internal_state)
  File "/Users/runner/runners/2.165.0/work/1/s/numba/core/compiler_machinery.py", line 264, in check
    mangled = func(compiler_state)
  File "/Users/runner/runners/2.165.0/work/1/s/numba/core/typed_passes.py", line 96, in run_pass
    raise_errors=self._raise_errors)
  File "/Users/runner/runners/2.165.0/work/1/s/numba/core/typed_passes.py", line 68, in type_inference_stage
    infer.propagate(raise_errors=raise_errors)
  File "/Users/runner/runners/2.165.0/work/1/s/numba/core/typeinfer.py", line 985, in propagate
    raise errors[0]
numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
non-precise type pyobject
[1] During: typing of argument at /Users/runner/runners/2.165.0/work/1/s/numba/tests/npyufunc/test_errors.py (28)

File "numba/tests/npyufunc/test_errors.py", line 28:
def floordiv(a, b):
    return a // b
    ^
```

**Comment by eric-wieser** [comment]
CI fixed :)

File: numba/np/numpy_support.py [hunk 2]
```diff
@@ -189,10 +193,6 @@ def _check_struct_alignment(rec, fields):
                 raise ValueError(msg.format(npy_align, llvm_align, dt))
 
 
-def is_arrayscalar(val):
-    return np.dtype(type(val)) in FROM_DTYPE
-
-
```
Comment (eric-wieser) [inline numba/np/numpy_support.py]: This wasn't really a useful function - seeminly it either meant:

* Actually an array scalar type, `isinstance(val, np.scalar)`
* Can be coerced into a array, which just covers all types (`SomeRandomType` is treated as `np.object`).

Comment (stuartarchibald) [reply numba/np/numpy_support.py]: I agree with your assessment of the function purpose. Seems like this is only used internally anyway so its removal is fine if required (I had a quick check on github search, and also looked in https://github.com/numba/numba/issues/5214#issuecomment-593611360).

**Pull request merged**